src/modules/keycloak-handler/keycloack-service-provider.event-handler.ts

```diff
@@ -0,0 +1,93 @@
+import { Injectable } from '@nestjs/common';
+import { ServiceProviderRepo } from '../service-provider/repo/service-provider.repo.js';
+
+import { RolleServiceProviderEntity } from '../rolle/entity/rolle-service-provider.entity.js';
+import {
+    PersonenkontextUpdatedData,
+    PersonenkontextUpdatedEvent,
+} from '../../shared/events/personenkontext-updated.event.js';
+import { EventHandler } from '../../core/eventbus/decorators/event-handler.decorator.js';
+import { KeycloakUserService } from '../keycloak-administration/index.js';
+import { RolleID } from '../../shared/types/aggregate-ids.types.js';
+
+//import { ClassLogger } from '../../../core/logging/class-logger.js';
+
+export type KontextIdsAndDuplicationFlag = {
+    hasDuplicateRolleIds: boolean;
+    personenkontextIdSet: Set<string>;
+};
+export type KeycloakRole = string;
+@Injectable()
+export class KeycloackServiceProviderHandler {
+    public constructor(
+        //private readonly logger: ClassLogger,
+        private readonly serviceRepo: ServiceProviderRepo,
+        private readonly KeycloackService: KeycloakUserService,
+    ) {}
+
+    public async fetchFilteredRolesDifference(
+        currentRoles: RolleID | string[],
+        changingRole: RolleID | string[],
+    ): Promise<(KeycloakRole | undefined)[]> {
+        const allRolleServiceProviders: RolleServiceProviderEntity[] =
+            await this.serviceRepo.fetchRolleServiceProvidersWithoutPerson(changingRole);
+
+        const specificRolleServiceProviders: RolleServiceProviderEntity[] =
+            await this.serviceRepo.fetchRolleServiceProvidersWithoutPerson(currentRoles);
+
+        const allServiceProvidersNames: Set<KeycloakRole | undefined> = new Set(
+            allRolleServiceProviders.map((element: RolleServiceProviderEntity) => element.serviceProvider.keycloakRole),
+        );
+
+        const specificServiceProvidersNames: Set<KeycloakRole | undefined> = new Set(
+            specificRolleServiceProviders.map(
+                (element: RolleServiceProviderEntity) => element.serviceProvider.keycloakRole,
+            ),
+        );
+
+        const updateRole: (KeycloakRole | undefined)[] = Array.from(allServiceProvidersNames).filter(
+            (role: KeycloakRole | undefined) => !specificServiceProvidersNames.has(role),
+        );
+
+        return updateRole;
+    }
+
+    @EventHandler(PersonenkontextUpdatedEvent)
+    public async updatePersonenkontexteOrDeleteKCandSP(event: PersonenkontextUpdatedEvent): Promise<void> {
+        const { newKontexte, currentKontexte, removedKontexte, person }: PersonenkontextUpdatedEvent = event;
+        const newRolle: RolleID | undefined = newKontexte?.[0]?.rolleId;
```

Comment:
new, deleted and current contexts are arrays. You can't use the first one only.
Also consider that a Teacher can change the school. We will have a new context with role teacher and a deleted with a role teacher, but for keycloak it does not change.